### PR TITLE
Add Pix2Pix model

### LINKED ISSUE #512: Add Pix2Pix
## 🚀 Feature
Pix2Pix is a really good example of image-to-image translation network and should be included in Lightning.

### Motivation

Currently Lightning doesn't have Pix2Pix model implemented. As it is a good starting example for anyone learning GANs, it should be easily accessible for training and experimenting with custom dataset.

### Pitch

I would like to take the responsibility of implementing Pix2Pix in Lightning.

### Alternatives

Many good examples are already available in PyTorch. For example- https://github.com/mrzhu-cool/pix2pix-pytorch

### PR DESCRIPTION
**Pull request by aniketmaurya**
## What does this PR do?
Fixes #512 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_bolts/models/gans/pix2pix/components.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+import torch
+from torch import nn
+
+
+def center_crop(image, new_shape):
```
Comment (oke-aditya) [inline pl_bolts/models/gans/pix2pix/components.py:5]: I think torchvision transforms has `center_crop`, if possible ew can use that perhaps?

File: pl_bolts/models/gans/pix2pix/components.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    h, w = image.shape[-2:]
+    n_h, n_w = new_shape[-2:]
+    cy, cx = int(h / 2), int(w / 2)
+    xmin, ymin = cx - n_w // 2, cy - n_h // 2
+    xmax, ymax = xmin + n_w, ymin + n_h
+    cropped_image = image[..., xmin:xmax, ymin:ymax]
+    return cropped_image
+
+
+class ConvBlock(nn.Module):
```
Comment (oke-aditya) [inline pl_bolts/models/gans/pix2pix/components.py:15]: I see that this block is really very frequently used. Conv + Bn + Act or Conv + Act + Bn.

Marking for #427 discussion.

File: pl_bolts/models/gans/pix2pix/components.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            x = self.dropout(x)
+        x = self.activation(x)
+        return x
+
+
+class UpSampleConv(nn.Module):
+
+    def __init__(self, input_channels, use_dropout=False, use_bn=True):
+        super(self).__init__()
+        self.upsample = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)
```
Comment (oke-aditya) [inline pl_bolts/models/gans/pix2pix/components.py:43]: Let's just confirm once if Upsample is done using `nn.Upsample` or `nn.ConvTranspose2d` both work fine. I haven't read Pix2Pix paper so let me check once.

File: pl_bolts/models/gans/pix2pix/components.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            print(x.shape)
+        x = self.conv_final(x)
+
+        return self.sigmoid(x)
+
+
+class Discriminator(nn.Module):
+
+    def __init__(self, input_channels, hidden_channels=8):
+        super(Discriminator, self).__init__()
```
Comment (oke-aditya) [inline pl_bolts/models/gans/pix2pix/components.py:151]:
```suggestion
        `super().__init__()`
```

Either work fine, though.

**Review by oke-aditya — COMMENTED** [review]
I had glance of it, let me know what you think.
Very nice PR, Pix2Pix is really useful model.

File: pl_bolts/models/gans/pix2pix/components.py (hunk 1 above)
Comment (akihironitta) [reply pl_bolts/models/gans/pix2pix/components.py:5]: In my opinion, we already have Bolts heavily depend on torchvision, so we could just use torchvision's one.

@Broda In general, should we have our own implementation in Bolts instead of employing one from major open-source libraries?

File: pl_bolts/models/gans/pix2pix/components.py (hunk 3 above)
Comment (aniketmaurya) [reply pl_bolts/models/gans/pix2pix/components.py:43]: Hi, thank you for the review. In section 6 of the Pix2Pix paper authors have mentioned that they upsampled the tensors by a factor of 2 but they haven't exactly mentioned if Transposed Conv is used or Upsample followed by Conv layer.

File: pl_bolts/models/gans/pix2pix/components.py (hunk 1 above)
Comment (aniketmaurya) [reply pl_bolts/models/gans/pix2pix/components.py:5]: Thank you for the suggestion! Will make the relevant change.

Comment (oke-aditya) [reply pl_bolts/models/gans/pix2pix/components.py:5]: > @Broda In general, should we have our own implementation in Bolts instead of employing one from major open-source libraries?

Sorry @Broda wrong ping 😅

It's for @Borda

Comment (akihironitta) [reply pl_bolts/models/gans/pix2pix/components.py:5]: @Broda I'm very sorry for the wrong ping.

File: pl_bolts/models/gans/__init__.py [hunk 5]
```diff
@@ -1,7 +1,9 @@
 from pl_bolts.models.gans.basic.basic_gan_module import GAN  # noqa: F401
 from pl_bolts.models.gans.dcgan.dcgan_module import DCGAN  # noqa: F401
+from pl_bolts.models.gans.pix2pix.pix2pix_module import Pix2Pix
```
Comment (akihironitta) [inline pl_bolts/models/gans/__init__.py:3]: Not related to this PR, but could you include this suggestion as well?
```suggestion
from pl_bolts.models.gans.basic.basic_gan_module import GAN
from pl_bolts.models.gans.dcgan.dcgan_module import DCGAN
from pl_bolts.models.gans.pix2pix.pix2pix_module import Pix2Pix
```

File: pl_bolts/models/gans/__init__.py [hunk 6]
```diff
@@ -1,7 +1,9 @@
...
 
 __all__ = [
     "GAN",
     "DCGAN",
+    "Pix2Pix"
```
Comment (akihironitta) [inline pl_bolts/models/gans/__init__.py:8]: Let's keep the comma at the end for the consistent `yapf` format.
```suggestion
    "Pix2Pix",
```

**Review by akihironitta — COMMENTED** [review]
Thank you for your work! Could you have a look at my nitpicking comments? :sweat_smile:

File: pl_bolts/models/gans/__init__.py (hunk 6 above)
Comment (aniketmaurya) [reply pl_bolts/models/gans/__init__.py]: I have done the mentioned change.

Comment (akihironitta) [reply pl_bolts/models/gans/__init__.py]: Thank you for your swift action!

File: pl_bolts/models/gans/pix2pix/components.py (hunk 1 above)
Comment (aniketmaurya) [reply pl_bolts/models/gans/pix2pix/components.py:5]: Hi @oke-aditya @akihironitta, `PyTorch CenterCrop` requires to pass the desired output size before creating the object. I will have to instantiate the `transforms.CenterCrop` on each Upsample block. Please, let me know what do you think.

Comment (aniketmaurya) [reply pl_bolts/models/gans/pix2pix/components.py:5]: Ok, I didn't know there is a functional `center_crop` exists in torchvision. It should work.

**Comment by aniketmaurya** [comment]
Hi I have added the training step code to the PR. Also I have tried to [train the model using Facades dataset](https://github.com/aniketmaurya/pytorch-gans/blob/main/pix2pix/Pix2Pix_pytorch.ipynb).
Please review and verify the model.

File: pl_bolts/models/gans/pix2pix/components.py (hunk 3 above)
Comment (akihironitta) [reply pl_bolts/models/gans/pix2pix/components.py:43]: I quickly checked [the paper](https://arxiv.org/pdf/1611.07004.pdf) and found that [the PyTorch implementation](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/f13aab8148bd5f15b9eb47b690496df8dadbab0c/models/networks.py#L315) linked from [the author's Lua implementation](https://github.com/phillipi/pix2pix) uses `nn.ConvTranspose2d`, so shall we follow that architecture unless someone has a strong opinion?

File: pl_bolts/models/gans/pix2pix/components.py (hunk 1 above)
Comment (akihironitta) [reply pl_bolts/models/gans/pix2pix/components.py]: > I think torchvision transforms has `center_crop`, if possible we can use that perhaps?
> Or maybe the idea is to avoid depdence on torchvision ?

@oke-aditya I think we should rely on very major libraries like torchvision instead of having our own implementation because 1) they are (should be) well maintained and tested, and 2) Bolts has already heavy dependency on torchvision anyway. So, in this case, we should use [`torchvision.transforms.functional.center_crop`](https://pytorch.org/vision/stable/transforms.html#torchvision.transforms.functional.center_crop). What's your thought on this?

**Review by akihironitta — COMMENTED** [review]
@aniketmaurya Would you mind merging master and applying yapf to pass the failing tests?

File: pl_bolts/models/gans/pix2pix/components.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    def __init__(self, in_channels, use_dropout=False, use_bn=False):
+        super().__init__()
+        self.maxpool = nn.MaxPool2d(kernel_size=2, stride=2)
+
+        if use_bn:
+            self.batchnorm = nn.BatchNorm2d(in_channels * 2)
+        self.use_bn = use_bn
+
+        if use_dropout:
+            self.dropout = nn.Dropout()
```
Comment (oke-aditya) [inline pl_bolts/models/gans/pix2pix/components.py:87]: Probably we need a probablitiy for dropout say 0.3 ?

File: pl_bolts/models/gans/pix2pix/components.py (hunk 3 above)
Comment (oke-aditya) [reply pl_bolts/models/gans/pix2pix/components.py:43]: Yes I too confirmed that it is `nn.ConvTranspose2d`. I have referred [TensorFlow docs,](https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/generative/pix2pix.ipynb#scrollTo=nhgDsHClSQzP) which give a really nice implementation.

File: pl_bolts/models/gans/pix2pix/components.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        self.upsample = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)
+        self.conv1 = nn.Conv2d(input_channels, input_channels // 2, kernel_size=2)
+        self.conv2 = nn.Conv2d(input_channels, input_channels // 2, kernel_size=3, padding=1)
+        self.conv3 = nn.Conv2d(input_channels // 2, input_channels // 2, kernel_size=2, padding=1)
+        if use_bn:
+            self.batchnorm = nn.BatchNorm2d(input_channels // 2)
+        self.use_bn = use_bn
+        self.activation = nn.ReLU()
+        if use_dropout:
+            self.dropout = nn.Dropout()
```
Comment (oke-aditya) [inline pl_bolts/models/gans/pix2pix/components.py:52]: Probably we need probablility of `0.5` here ? as from docs.

**Review by oke-aditya — COMMENTED** [review]
I'm bit unsure about some aspects of implementation.
[This colab notebook ](https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/generative/pix2pix.ipynb#scrollTo=nhgDsHClSQzP) gives a really simple and effective Pix2Pix.

**Comment by oke-aditya** [comment]
GANs are notorius :sweat_smile:  Small differences such as LeakyRelu vs Relu, Dropout probability, ConvTranspose vs Upsampling really make difference. Really tough time in getting a GAN trained, really great job @aniketmaurya 

@akihironitta I would suggest to use torchvision, it is well tested and a standard (over 15Million downloads).
Even with lightning flash we use it now :smile: It avoids lot of code writing.

**Comment by akihironitta** [comment]
@oke-aditya I agree, thanks for your comment :)

File: requirements.txt [hunk 9]
```diff
@@ -1,2 +1,3 @@
 torch>=1.6
-pytorch-lightning>=1.1.1, <1.2
+torchvision>=0.8
```
Comment (akihironitta) [inline requirements.txt:2]: `torchvision` is currently an optional package for Bolts, so would you include the change in `requirements/models.txt`?:
https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/86e3f52308fa02e2a988b4977858a945f7d15ab2/requirements/models.txt#L1

**Review by akihironitta — COMMENTED** [review]
Hey @aniketmaurya, it'd be great to have this model in Bolts!!! Please mark this PR as ready for review when you're ready :)

**Review by Borda — APPROVED** [review]

**Comment by emoebel** [comment]
I'm trying to use `pl_bolts.models.gans.Pix2Pix`, and I am able to run the training. However, there seems to be a problem with logging and checkpointing. I am using tensorboard to view the metrics (with `TensorBoardLogger`), and everything works fine when I use `pl_bolts.models.gans.GAN` and `pl_bolts.models.gans.DCGAN`. I use the same script with `Pix2Pix`, and no checkpoints are saved, and nothing is logged to tensorboard. Any idea?
